### PR TITLE
Kerberos authentication method added.

### PR DESCRIPTION
**Pull request by kwart**
OS part of Kerberos authentication.
EE PR: hazelcast/hazelcast-enterprise#3574

This PR introduces Kerberos ticket authentication support into Hazelcast. This OS part only helps with the configuration. The main Kerberos related logic is part of the Hazelcast Enterprise.

Main configuration changes:
* added 'kerberos' identity (clients & members)
* added 'kerberos' authentication (members)
* added limited security `realm` configuration to client-side - allows specifying named JAAS authentication

Implementation:
* added new credentials factory, which is able to retrieve a Kerberos service ticket: `com.hazelcast.security.impl.KerberosCredentialsFactory`
* added new login module, which accepts/verifies the Kerberos service tickets: `com.hazelcast.security.loginimpl.GssApiLoginModule`

### Kerberos Identity

The Kerberos identity configures a credentials factory to provide Kerberos service tickets.

Default Service principal names for Hazelcast members are in the form `hz/address@REALM` (e.g. `hz/192.168.1.1@ACME.COM`). The `security-ream` configuration allows delegating Kerberos authentication to another realm configured within the Hazelcast configuration.

```xml
<realm name="kerberosRealm">
    <identity>
        <kerberos>
            <realm>ACME.COM</realm>
            <security-realm>krb5Initiator</security-realm>
        </kerberos>
    </identity>
</realm>
<realm name="krb5Initiator">
    <authentication>
        <jaas>
            <login-module class-name="com.sun.security.auth.module.Krb5LoginModule" usage="REQUIRED">
                <properties>
                    <property name="useTicketCache">true</property>
                    <property name="doNotPrompt">true</property>
                </properties>
            </login-module>
        </jaas>
    </authentication>
</realm>
```

#### Properties of Kerberos identity

The `KerberosCredentialsFactory` has the following properties, which can be specified in the configuration:

| Property name | Default value | Description |
|---|---|---|
|spn| |Allows configuring static service principal name (SPN). It's meant for use-cases where all members share a single Kerberos identity.|
|serviceNamePrefix| `"hz/"` | Defines prefix of the Service Principal Name. By default the member's principal name (for which this credentials factory asks the service ticket) is in form `"[servicePrefix][memberIpAddress]@[REALM]"` (e.g. `"hz/192.168.1.1@ACME.COM"`).|
| realm| | Kerberos realm name (e.g. `"ACME.COM"`)
| securityRealm | | Security realm name in Hazelcast configuration used for Kerberos authentication. The authentication configuration in the referenced security realm will be used to fill the Subject with Kerberos credentials (e.g. TGT). |

### Kerberos Authentication

The corresponding authenticating part on the server-side is able to accept the Kerberos tickets and verify them. Again the Kerberos authentication is delegated to another realm with Kerberos login module configured.

```xml
<realm name="kerberosRealm">
    <authentication>
        <kerberos>
            <security-realm>krb5Acceptor</security-realm>
        </kerberos>
    </authentication>
</realm>
<realm name="krb5Acceptor">
    <authentication>
        <jaas>
            <login-module class-name="com.sun.security.auth.module.Krb5LoginModule" usage="REQUIRED">
                <properties>
                    <property name="isInitiator">false</property>
                    <property name="useTicketCache">false</property>
                    <property name="doNotPrompt">true</property>
                    <property name="useKeyTab">true</property>
                    <property name="storeKey">true</property>
                    <property name="principal">hz/192.168.1.1@ACME.COM</property>
                    <property name="keyTab">/opt/member1.keytab</property>
                </properties>
            </login-module>
        </jaas>
    </authentication>
</realm>
```

This configuration only loads Kerberos secrets from the keytab file and it doesn't authenticate against the Kerberos KDC.

#### Properties of Kerberos authentication

The `GssApiLoginModule` derives from the abstract `ClusterLoginModule` so it supports its options (`skipIdentity`, `skipEndpoint` and `skipRole`). It introduces also its own options:

| Property name | Default value | Description |
|---|---|---|
|relaxFlagsCheck| `false` | Allows disabling some of the checks on the incoming token (e.g. passes authentication even if the mutual authentication is required by the token).|
| securityRealm | | Security realm name in Hazelcast configuration used for Kerberos authentication. The authentication configuration in the referenced security realm will be used to fill the Subject with Kerberos credentials (e.g. Keytab). |

#### LDAP integration
The Kerberos authentication allows loading role mapping information from an LDAP server (usually the one backing the Kerberos KDC server too).

```xml
<realm name="kerberosRealm">
    <authentication>
        <kerberos>
            <skip-role>true</skip-role>
            <security-realm>krb5Acceptor</security-realm>
            <ldap>
                <url>ldap://127.0.0.1/</url>
                <system-user-dn>uid=admin,ou=system</system-user-dn>
                <system-user-password>secret</system-user-password>
                <skip-authentication>true</skip-authentication>
                <user-filter>(krb5PrincipalName={login})</user-filter>
                <role-mapping-attribute>cn</role-mapping-attribute>
            </ldap>
        </kerberos>
    </authentication>
</realm>
```

#### New Property in LdapLoginModule

LDAP login modules in Hazelcast were used to verify the credentials till now. Integrating it with Kerberos we don't want to verify (again) the credentials, but rather load roles for the Kerberos identity.
Therefore a new option is introduced, which allows proceeding without credentials verification:

| Property name | Default value | Description |
|---|---|---|
| skipAuthentication | `false` | Allows disabling password verification and only takes care about filling `HazelcastPrincipal` instances into the Subject.|

### Security realms on client-side

Security realms configuration was introduced on the client-side too. It allows specifying JAAS login modules which can be referenced from Kerberos identity configuration.

```xml
<security>
    <kerberos>
        <realm>ACME.COM</realm>
        <security-realm>krb5Initiator</security-realm>
    </kerberos>
    <realms>
        <realm name="krb5Initiator">
            <authentication>
                <jaas>
                    <login-module class-name="com.sun.security.auth.module.Krb5LoginModule" usage="REQUIRED">
                        <properties>
                            <property name="isInitiator">true</property>
                            <property name="useTicketCache">false</property>
                            <property name="doNotPrompt">true</property>
                            <property name="useKeyTab">true</property>
                            <property name="storeKey">true</property>
                            <property name="principal">jduke@HAZELCAST.COM</property>
                            <property name="keyTab">/opt/jduke.keytab</property>
                        </properties>
                    </login-module>
                </jaas>
            </authentication>
        </realm>
    </realms>
</security>
```

### Hazelcast security realms and JGSS

It's also possible to use a direct JGSS approach, where we don't specify 2 extra security realms in Hazelcast configuration. Users in such case provide a path to the JAAS login configuration file as the `-Djava.security.auth.login.config` argument. The login configurations have to be named according to JGSS implementation naming:

```
com.sun.security.jgss.initiate {
  com.sun.security.auth.module.Krb5LoginModule required
  useTicketCache=true
  doNotPrompt=true
  debug=true
  ;
};
 
com.sun.security.jgss.accept {
  com.sun.security.auth.module.Krb5LoginModule required
  storeKey=true
  doNotPrompt=true
  debug=true
  useKeyTab=true
  keyTab="c:/hazelcast/member1.keytab"
  principal="hazelcast/member1.acme.com@ACME.COM"
  isInitiator=false
  ;
};
```

### Limitations
* This Kerberos authentication is only able to validate connections on the server-side. It doesn't support mutual authentication.
* The GSS-API is not used for protecting (wrapping) messages after authentication (e.g. encryption, integrity checks). It's only used for accepting tokens.
* The token itself is not protected against Man-in-the-Middle attacks. If an attacker is able to eavesdrop the token and use it before the original sender, then the attacker succeeds with the authentication but the original sender won't.
  * There is replay protection in Java which caches already used tokens.
  * Java Kerberos implementation accepts the token for 5min  (by default) from its creation.
* Time has to be synchronized on the machines where the Kerberos is used.

If you are running Hazelcast in an untrusted network with a MITM attack risk, then enable encryption on Hazelcast protocols to prevent stealing the token.

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/48/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/48/Hazelcast-pr-builder-48.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/66/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/66/Hazelcast-pr-builder-66.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/MapStore.java:90: Line has trailing spaces. [RegexpSingleline]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/MapStore.java:91: Line has trailing spaces. [RegexpSingleline]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/67/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/67/Hazelcast-pr-builder-67.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain 
information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/MapStore.java:91: Line has trailing spaces. [RegexpSingleline]
--------------------------
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/config/security/LdapAuthenticationConfig.java [hunk 1]
```diff
@@ -228,85 +232,85 @@ public LdapAuthenticationConfig setUserSearchScope(LdapSearchScope userSearchSco
...
 
         return new LoginModuleConfig[] { loginModuleConfig };
     }
 
     @Override
     public String toString() {
-        return "LdapAuthenticationConfig [url=" + url + ", socketFactoryClassName=" + socketFactoryClassName + ", systemUserDN="
-                + systemUserDn + ", systemUserPassword=***, parseDN=" + parseDn + ", roleContext="
+        return "LdapAuthenticationConfig [url=" + url + ", socketFactoryClassName=" + socketFactoryClassName + ", systemUserDn="
+                + systemUserDn + ", systemUserPassword=" + systemUserPassword + ", parseDn=" + parseDn + ", roleContext="
```
Comment (galibey) [inline hazelcast/src/main/java/com/hazelcast/config/security/LdapAuthenticationConfig.java:264]: Why do we print a secret `systemUserPassword` here?

Comment (kwart) [reply hazelcast/src/main/java/com/hazelcast/config/security/LdapAuthenticationConfig.java:264]: Nice catch. I'll fix it. It was caused by regenerating the `toString` method in the Eclipse IDE.

**Comment by kwart** [comment]
Thank you for the reviews, Alexander, Stephan, Sancar!

@sancar and @zpqrtbnk, could you also have a look into the related Enterprise PR hazelcast/hazelcast-enterprise#3574?

**Comment by zpqrtbnk** [comment]
Test: run 4.1-SNAPSHOT server with this configuration:
```
<realms>
      <realm name="kerberosRealm">
        <authentication>
          <kerberos>
            <security-realm>krb5Acceptor</security-realm>
          </kerberos>
        </authentication>
      </realm>
      <realm name="krb5Acceptor">
        <authentication>
          <jaas>
            <login-module class-name="com.sun.security.auth.module.Krb5LoginModule" usage="REQUIRED">
              <properties>
                <property name="isInitiator">false</property>
                <property name="useTicketCache">false</property>
                <property name="doNotPrompt">true</property>
                <property name="useKeyTab">true</property>
                <property name="storeKey">true</property>
                <property name="principal">hz/cluster1234@HZ.LOCAL</property>
                <property name="keyTab">/d/d/Hazelcast/CSharpClient_v4/krb5/hzcluster1234.keytab</property>

                <!-- relax flags check because .NET tokens have too many things -->
                <property name="relaxFlagsCheck">true</property>
              </properties>
            </login-module>
          </jaas>
        </authentication>
      </realm>
    </realms>
    <client-authentication realm="kerberosRealm"/>
```

And then run the [KerberosTests](https://github.com/hazelcast/hazelcast-csharp-client/blob/master/Hazelcast.Test/Hazelcast.Client.Test/KerberosTests.cs) of the .NET client. The test **FAILS** with the following error being reported on the server:

```
avr. 21, 2020 2:03:58 PM com.hazelcast.client.impl.protocol.task.AuthenticationCustomCredentialsMessageTask
WARNING: [127.0.0.1]:5701 [c0ee2b1e-6e1b-4bc1-bb69-b19f63a8d715] [4.1-SNAPSHOT] Received auth from Connection[id=30, /127.0.0.1:5701->/127.0.0.1:21390, qualifier=null, endpoint=null, alive=true, connectionType=NONE] with clientUuid 34a619bd-f75b-44cd-9295-bc577ceae647, authentication failed
avr. 21, 2020 2:03:58 PM com.hazelcast.client.impl.protocol.task.AuthenticationCustomCredentialsMessageTask
WARNING: [127.0.0.1]:5701 [c0ee2b1e-6e1b-4bc1-bb69-b19f63a8d715] [4.1-SNAPSHOT] Accepting the GSS-API token failed. Failure unspecified at GSS-API level (Mechanism level: Invalid argument (400) - Cannot find key of appropriate type to decrypt AP REP - RC4 with HMAC)
javax.security.auth.login.LoginException: Accepting the GSS-API token failed. Failure unspecified at GSS-API level (Mechanism level: Invalid argument (400) - Cannot find key of appropriate type to decrypt AP REP - RC4 with HMAC)
        at com.hazelcast.security.loginimpl.GssApiLoginModule.acceptToken(GssApiLoginModule.java:107)
        at com.hazelcast.security.loginimpl.GssApiLoginModule.lambda$onLogin$0(GssApiLoginModule.java:68)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Unknown Source)
        at com.hazelcast.security.loginimpl.GssApiLoginModule.onLogin(GssApiLoginModule.java:68)
        at com.hazelcast.security.ClusterLoginModule.login(ClusterLoginModule.java:68)
        at sun.reflect.GeneratedMethodAccessor11.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at javax.security.auth.login.LoginContext.invoke(Unknown Source)
        at javax.security.auth.login.LoginContext.access$000(Unknown Source)
        at javax.security.auth.login.LoginContext$4.run(Unknown Source)
        at javax.security.auth.login.LoginContext$4.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.login.LoginContext.invokePriv(Unknown Source)
        at javax.security.auth.login.LoginContext.login(Unknown Source)
        at com.hazelcast.client.impl.protocol.task.AuthenticationBaseMessageTask.authenticate(AuthenticationBaseMessageTask.java:131)
        at com.hazelcast.client.impl.protocol.task.AuthenticationBaseMessageTask.authenticate(AuthenticationBaseMessageTask.java:114)
        at com.hazelcast.client.impl.protocol.task.AuthenticationBaseMessageTask.processMessage(AuthenticationBaseMessageTask.java:82)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:153)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:116)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
        at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
        at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

i.e. what seems to be an encryption configuration issue. Investigating.

**Comment by zpqrtbnk** [comment]
Update: above error can be caused by a missing `keytab` file. Replacing the path `/d/d/Hazelcast/CSharpClient_v4/krb5/hzcluster1234.keytab` in the configuration with the more Windows-friendly `d:\d\Hazelcast\CSharpClient_v4\krb5\hzcluster1234.keytab` (server runs on Windows) indeed leads to a different error:

```
WARNING: [127.0.0.1]:5701 [418fb672-bf92-4897-9df1-f623ed9e2841] [4.1-SNAPSHOT] Received auth from Connection[id=30, /127.0.0.1:5701->/127.0.0.1:21511, qualifier=null, endpoint=null, alive=true, connectionType=NONE] with clientUuid 0bb4e818-6540-4bfd-a8c7-50999b8af040, authentication failed
avr. 21, 2020 2:13:30 PM com.hazelcast.client.impl.protocol.task.AuthenticationCustomCredentialsMessageTask
WARNING: [127.0.0.1]:5701 [418fb672-bf92-4897-9df1-f623ed9e2841] [4.1-SNAPSHOT] Confidentiality and data integrity is not provided by this login module.
javax.security.auth.login.FailedLoginException: Confidentiality and data integrity is not provided by this login module.
        at com.hazelcast.security.loginimpl.GssApiLoginModule.acceptToken(GssApiLoginModule.java:98)
        at com.hazelcast.security.loginimpl.GssApiLoginModule.lambda$onLogin$0(GssApiLoginModule.java:68)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Unknown Source)
        at com.hazelcast.security.loginimpl.GssApiLoginModule.onLogin(GssApiLoginModule.java:68)
        at com.hazelcast.security.ClusterLoginModule.login(ClusterLoginModule.java:68)
        at sun.reflect.GeneratedMethodAccessor11.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at javax.security.auth.login.LoginContext.invoke(Unknown Source)
        at javax.security.auth.login.LoginContext.access$000(Unknown Source)
        at javax.security.auth.login.LoginContext$4.run(Unknown Source)
        at javax.security.auth.login.LoginContext$4.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.login.LoginContext.invokePriv(Unknown Source)
        at javax.security.auth.login.LoginContext.login(Unknown Source)
        at com.hazelcast.client.impl.protocol.task.AuthenticationBaseMessageTask.authenticate(AuthenticationBaseMessageTask.java:131)
        at com.hazelcast.client.impl.protocol.task.AuthenticationBaseMessageTask.authenticate(AuthenticationBaseMessageTask.java:114)
        at com.hazelcast.client.impl.protocol.task.AuthenticationBaseMessageTask.processMessage(AuthenticationBaseMessageTask.java:82)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:153)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:116)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
        at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
        at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```
Investigating...

**Comment by kwart** [comment]
@zpqrtbnk It's probably the misplaced flag for relaxing the checks. It should be part of the `<kerberos>` authentication.

```xml
<realms>
      <realm name="kerberosRealm">
        <authentication>
          <kerberos>
            <relax-flags-check>true</relax-flags-check>
            <security-realm>krb5Acceptor</security-realm>
          </kerberos>
        </authentication>
      </realm>
      <realm name="krb5Acceptor">
        <authentication>
          <jaas>
            <login-module class-name="com.sun.security.auth.module.Krb5LoginModule" usage="REQUIRED">
              <properties>
                <property name="isInitiator">false</property>
                <property name="useTicketCache">false</property>
                <property name="doNotPrompt">true</property>
                <property name="useKeyTab">true</property>
                <property name="storeKey">true</property>
                <property name="principal">hz/cluster1234@HZ.LOCAL</property>
                <property name="keyTab">/d/d/Hazelcast/CSharpClient_v4/krb5/hzcluster1234.keytab</property>
              </properties>
            </login-module>
          </jaas>
        </authentication>
      </realm>
    </realms>
```

**Comment by zpqrtbnk** [comment]
Thanks - with this fix, the test passes with a hard-coded `<all-permissions principal="hzclient@HZ.LOCAL" />` permission. Now testing with LDAP.

**Comment by zpqrtbnk** [comment]
Took a bit of time to figure out LDAP but the following configuration works (copying here for reference):

```
<ldap>
  <url>ldap://server19.hz.local/</url>

  <system-user-dn>CN=Administrateur,CN=Users,DC=hz,DC=local</system-user-dn>
  <system-user-password>****</system-user-password>

  <skip-authentication>true</skip-authentication>

  <user-context>CN=Users,DC=hz,DC=local</user-context>
  <user-search-scope>subtree</user-search-scope>
  <user-filter>(userPrincipalName={login})</user-filter>

  <role-mapping-mode>reverse</role-mapping-mode>
  <role-context>CN=Users,DC=hz,DC=local</role-context>
  <role-search-scope>subtree</role-search-scope>
  <role-filter>(member={memberDN})</role-filter>
  <role-recursion-max-depth>4</role-recursion-max-depth>
  <role-name-attribute>cn</role-name-attribute>
</ldap>
```

It rightfully finds that the user is in group `hzcluster1234users` and therefore the following permission works:

```
<all-permissions principal="hzcluster1234users" />
```

So... all good on my side!

**Comment by kwart** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/200/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/200/Hazelcast-pr-builder-200.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientJmxMetricsTest.testNoMBeanLeak:52 expected:<0> but was:<13>
[INFO] 
[ERROR] Tests run: 3854, Failures: 1, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by kwart** [comment]
Unrelated test failure (#16786).

**Pull request merged**